peekaboo/daemon.py

```diff
@@ -126,40 +127,57 @@ class PeekabooStreamRequestHandler(SocketServer.StreamRequestHandler):
     """
     def handle(self):
         """
-        Handles a analysis request. The path of the directory / file to analyse must
-        be written to the corresponding socket.
-        The maximum buffer size is 1024 bytes.
+        Handles an analysis request. This is expected to be a JSON structure
+        containing the path of the directory / file to analyse.
+        The maximum buffer size is 16 KiB, because JSON incurs some bloat.
         """
         self.request.sendall('Hallo das ist Peekaboo\n\n')
-        path = self.request.recv(1024).rstrip()
-        logger.info("Got run_analysis request for %s" % path)
-
-        if not os.path.exists(path):
-            self.request.sendall(
-                'FEHLER: Pfad existiert nicht oder Zugriff verweigert.'
-            )
-            logger.error("ERROR: Path does not exist or no permission to access it.")
-        else:
-            for_analysis = []
-            if os.path.isfile(path):
-                sample = make_sample(path, self.request)
-                if sample:
-                    for_analysis.append(sample)
-            else:
-                # walk recursively through entries in the given directory.
-                for dirname, __, filenames in os.walk(path):
-                    for filename in filenames:
-                        logger.debug("Found file %s" % filename)
-                        f = os.path.join(dirname, filename)
-                        sample = make_sample(f, self.request)
-                        if sample:
-                            for_analysis.append(sample)
-
-            # introduced after an issue where results were reported
-            # before all files could be added.
-            for sample in for_analysis:
-                ConnectionMap.add(self.request, sample)
-                JobQueue.submit(sample, self.__class__)
+        request = self.request.recv(1024 * 16).rstrip()
+
+        try:
+            parts = json.loads(request)
+        except:
+            self.request.sendall('FEHLER: Ungueltiges JSON.')
+            logger.error('Invalid JSON in request.')
+            return
+
+        if type(parts) is not dict:
+            self.request.sendall('FEHLER: Ungueltiges Datenformat.')
+            logger.error('Invalid data structure.')
+            return
+
+        for_analysis = []
+        for partname in parts:
+            part = parts[partname]
+
+            if not part.has_key('full_name'):
+                self.request.sendall('FEHLER: Unvollstaendige Datenstruktur.')
+                logger.error('Incomplete data structure.')
+                continue
```

Comment:
Should we continue analysing the parts we can get a hold (as it is now with continue) of or should we bail on the entire batch if any one part is fishy (return)? I guess, it won't work as of now anyway, because we report an error but continue on...

Reply:
In that case it makes sense to bail on the entire operation.
@Jack28 ?